var/spack/repos/builtin/packages/python/package.py

```diff
@@ -319,6 +334,144 @@ def filter_compilers(self, prefix):
     # ========================================================================
 
     @property
+    def executable(self):
+        """Returns the path to the Python executable, which may vary depending
+        on the version of Python and how it was installed.
+
+        In general, Python 2 comes with ``python`` and ``python2`` executables,
+        while Python 3 only comes with a ``python3`` executable.
+
+        :returns: The full path to the Python executable
+        :rtype: str
+        """
```

Comment:
Ok, I have some critiques of this:

1. I think the comment is not quite right about `python`.
2. I think the name `executable` is too generic.  Personally, I think this should just be called `python` for the python package (or something more specific -- but see below).
3. I am not sure that we enable packages to comply adequately with [PEP 394](https://www.python.org/dev/peps/pep-0394) if we just have a single `executable` function.

For (1), take a look at the [Recommendation](https://www.python.org/dev/peps/pep-0394/#recommendation) section of PEP 394:

> * The more general python command should be installed whenever any version of Python 2 is installed and should invoke the same version of Python as the python2 command (however, note that some distributions have already chosen to have python implement the python3 command; see the Rationale and Migration Notes below).

So, we probably shouldn't say that Python 3 comes with `python3` (even though it's usually true).

For (2) and (3), look at these recommendations, again from PEP 394:

> * In order to tolerate differences across platforms, all new code that needs to invoke the Python interpreter should not specify python , but rather should specify either python2 or python3 (or the more specific python2.x and python3.x versions; see the Migration Notes ). This distinction should be made in shebangs, when invoking from a shell script, when invoking via the system() call, or when invoking in any other context.
> * One exception to this is scripts that are deliberately written to be source compatible with both Python 2.x and 3.x. Such scripts may continue to use python on their shebang line without affecting their portability.

The way I interpret this, Python extensions need to be able to ask for either `python2`, `python3`, or `python`, depending on what they are compatible with.  I think these are the expectations from packages:

1. Packages that are only Python 2 compatible should ask for `spec['python'].python2`.
2. Packages that are only Python 3 compatibale should ask for `spec['python'].python3`
3. Packages that are single-source should ask for `spec['python'].python`

There are also dual-source packages, that install only the source for the version of Python they're built with, or that are compatible with only one version of Python depending on their version.  For that, you might want to be able to generically refer in a Spack package to the specific Python you're being built with.  So I would add *another* accessor for `pythonX`, as in the `X` in `pythonX.Y`:

* Packages that are dual-source or that need specific Python interpreters but for different versions of Python should ask for `spec['python'].pythonX`.

I think you have the logic for externals *mostly* correct in that it'll raise an error if you ask for `python2` from a Python 3 package or `python3` from a Python 2 package.  I think we shoud *also* raise an error if you ask for `python` from an external when `python` refers to some other installation's `python`.  e.g., on Arch linux, if you register the system Python as an external Python 2, and you ask for `python`, it's actually a Python 3 interpreter.  That should be an error.  Likewise if you register homebrew's Python 3 as an exeternal and ask for `python`, *that* should be an error if it's present, because it's actually a Python2 interpreter.

You can implement this by actually running `python -V` and making sure it matches your version.

Does that seem reasonable?  I think PEP 394 really does make things this complex (unfortunately).


Finally, this is off-topic, but look at this recommendation from PEP 394:

> * When reinvoking the interpreter from a Python script, querying `sys.executable` to avoid hardcoded assumptions regarding the interpreter location remains the preferred approach.

This has ramifications for @alalazo's `spack.environment.from_sourcing_files()`: it currently runs `python` but should be modified to run whatever `sys.executable` is.